// cdk-nextjs-standalone/lib/BundleFunction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundleFunction = exports.ESM_BUNDLE_DEFAULTS = void 0;
const fs = require("fs");
const os = require("os");
const path = require("path");
const path_1 = require("path");
const esbuild = require("esbuild");
exports.ESM_BUNDLE_DEFAULTS = {
    format: 'esm',
    mainFields: ['module', 'main'],
    banner: {
        // https://github.com/evanw/esbuild/issues/1921
        js: `import { createRequire } from 'module'; const require = createRequire(import.meta.url);`,
    },
};
/**
 * Compile a function handler with esbuild.
 * @returns bundle directory path
 */
function bundleFunction({ inputPath, outputPath, outputFilename, bundleOptions }) {
    if (!outputPath) {
        if (!outputFilename)
            outputFilename = bundleOptions.format === 'esm' ? 'index.mjs' : 'index.js';
        const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'nextjs-bundling-'));
        outputPath = path.join(tempDir, outputFilename);
    }
    const esbuildResult = esbuild.buildSync({
        ...bundleOptions,
        entryPoints: [inputPath],
        outfile: outputPath,
    });
    if (esbuildResult.errors.length > 0) {
        esbuildResult.errors.forEach((error) => console.error(error));
        throw new Error('There was a problem bundling the function.');
    }
    // console.debug('Bundled ', inputPath, 'to', outputPath);
    return path_1.dirname(outputPath);
}
exports.bundleFunction = bundleFunction;
